### ISSUE TITLE
Support LLVM-based Intel C/C++ compilers

### ISSUE THREAD

**Issue by KiriniriK** [microsoft/vscode-cpptools#10218]
### Environment

- OS and Version: Windows 11
- VS Code Version: 1.73.1
- C/C++ Extension Version: v1.12.4
- Other extensions you installed (and if the issue persists after disabling them):  CMake Tools v1.12.27

### Bug Summary and Steps to Reproduce

Bug Summary:
Cpptools uses weird args to query about the `icx` compiler, leading to fallbacks. Features such as template argument deduction are disabled. The workspace is a simple CMake project with a cpp file.

Steps to reproduce:
1. Open the project

### Expected behavior

No errors

### Code sample and Logs

The project contains two files, `CMakeLists.txt` and `main.cpp`.
CMake kit used:

```json
{
  "name": "IntelLLVM",
  "environmentSetupScript": "C:\\Program Files (x86)\\Intel\\oneAPI\\setvars.bat",
  "compilers": {
    "C": "C:\\Program Files (x86)\\Intel\\oneAPI\\compiler\\latest\\windows\\bin\\icx.exe",
    "CXX": "C:\\Program Files (x86)\\Intel\\oneAPI\\compiler\\latest\\windows\\bin\\icx.exe"
  },
  "preferredGenerator": {
    "name": "Ninja Multi-Config"
  }
}
```

`CMakeLists.txt`:
```cmake
cmake_minimum_required(VERSION 3.18 FATAL_ERROR)
project(vscodetest)

set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

add_executable(vscodetest main.cpp)

enable_testing()

add_test(NAME Test COMMAND vscodetest)
```

`main.cpp`:
```cpp
#include <random>

int main() {
    std::uniform_int_distribution a(1, 10);
    return 0;
}
```

This project is using the default Win32 config.
Language server log:
```
loggingLevel: Debug
Custom browse configuration received: {
  "browsePath": [
    "c:/users/abcdef/myproj/vscodetest"
  ],
  "compilerPath": "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe",
  "compilerArgs": [],
  "compilerFragments": [
    "/DWIN32 /D_WINDOWS /GR /EHsc /O2 /Ob2 /DNDEBUG -MD",
    "-Qstd=c++20"
  ]
}
```

```
cpptools/didChangeCustomBrowseConfiguration
Attempting to get defaults from C++ compiler in "compilerPath" property: 'c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe'
Querying compiler for default C++ language standard using command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -x c++ -E -dM nul
Querying compiler for default C++ language standard using command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -x c++ -E -dM /dev/null
Querying compiler for default C language standard using command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -x c -E -dM nul
Querying compiler for default C language standard using command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -x c -E -dM /dev/null
Querying compiler's default target using command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -dumpmachine
Unhandled default compiler target value detected: 
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=c++17 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=c++1z -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=c++14 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=c++1y -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=c++11 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=c++0x -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=c++03 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=c++98 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++23 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++2b -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++20 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++2a -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++17 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++1z -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++14 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++1y -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++11 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++0x -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++03 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -std=gnu++98 -m64 -Wp,-v -E -dM -x c++ nul
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -m64 -Wp,-v -E -dM -x c++ nul
Failed to query compiler. Falling back to 32-bit intelliSenseMode.
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -m32 -Wp,-v -E -dM -x c++ nul
Failed to query compiler. Falling back to no bitness.
Compiler query command line: "c:/program files (x86)/intel/oneapi/compiler/latest/windows/bin/icx.exe" -Qstd=c++20 -Wp,-v -E -dM -x c++ nul
Intel(R) oneAPI DPC++/C++ Compiler for applications running on Intel(R) 64, Version 2022.2.1 Build 20221101
Copyright (C) 1985-2022 Intel Corporation. All rights reserved.

icx: warning: unknown argument ignored in clang-cl: '-dM' [-Wunknown-argument]
icx: warning: unknown argument ignored in clang-cl: '-x' [-Wunknown-argument]
icx: error: no such file or directory: 'c++'
```

```
sending compilation args for C:\Users\abcdef\MyProj\vscodetest\main.cpp
  include: C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\ENTERPRISE\VC\TOOLS\MSVC\14.34.31933\INCLUDE
  include: C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\ENTERPRISE\VC\TOOLS\MSVC\14.34.31933\ATLMFC\INCLUDE
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.19041.0\UM
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.19041.0\UCRT
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.19041.0\SHARED
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.19041.0\WINRT
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.19041.0\CPPWINRT
  stdver: ms_c++14
  intelliSenseMode: windows-msvc-x64
```

### Screenshots

![image](https://user-images.githubusercontent.com/77284376/205199232-465664fd-4ebc-4da7-b368-b63875f6edaf.png)
`argument list for class template "std::uniform_int_distribution" is missing C/C++(441)`

**Comment by Colengms** [comment]
Hi @KiriniriK .  The C/C++ extension's compiler auto-configuration does not currently support the Intel compiler.  Currently, only cl.exe, gcc, clang, and compilers based on gcc or clang, are supported.  We could repurpose this issue as a Feature Request to add support for it.  It doesn't look like we have a Feature Request currently open specifically for the Intel compiler.

However, my understanding is that Intel is moving their compilers to LLVM/clang.  https://www.hpcwire.com/2021/08/10/intel-completes-llvm-adoption-will-end-updates-to-classic-c-c-compilers-in-future/

>“We recommend that all new projects start with the LLVM-based Intel C/C++ compilers, and all existing projects should make a plan to migrate to the new compiler this year. At some point in the future, the classic C/C++ compilers will enter “Legacy Product Support” mode signaling the end of regular updates to the classic compiler base, and they will no longer appear in oneAPI toolkits.”

**Comment by Colengms** [comment]
That said, now I realize that you are indeed using the LLVM-based compiler.  It looks like the error message you are seeing is technically coming from clang-cl, perhaps under the hood.  Clang-cl accepts CL command line args, not gcc/clang command line args.  So, it looks like we'll need to do something to correctly identify the Intel compiler and pass it the args it expects.

**Comment by Colengms** [comment]
There are various other oddities specific to the Intel LLVM compiler, such as the addition of "Qstd" (only on Windows) as an alternative to "std".  ([Command Line Options](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-dpcpp-cpp-compiler-dev-guide-and-reference/top/compiler-reference/compiler-options/alphabetical-option-list.html))

icx seems to accept gcc/clang args fine on Linux.  Only icx.exe on Windows requires clang-cl args.

**Comment by KiriniriK** [comment]
@Colengms Thanks for the reply. By the way, is there a way to retrieve only `browsePath` or `includePath` automatically and manually set `cStandard` or `cppStandard` for workaround?

**Comment by Colengms** [comment]
Hi @KiriniriK .  In your `c_cpp_properties.json` (run `C/C++: Edit Configuration (JSON)` if it doesn't exist yet), you can set the `compilerPath` set to `""`.  That will disable compiler querying (even when a custom configuration provider is set up and the field would otherwise have been ignored).

You can add system include paths that would otherwise have been retrieved from the compiler, to the `includePath` field of the configuration, to cause those files to be parsed.  However, IntelliSense would still use only the include paths specified by the custom configuration.  You may be able to work around that by specifying your system include paths in your CMakeLists.txt (i.e. with `target_include_directories`).  However, I don't know of a work-around for the standard version.

**Comment by KiriniriK** [comment]
@Colengms Thanks. I'm currently using `cl.exe` as `compilerPath` instead and disabled the custom configuration provider. The only inconvenience is that I have to supply include paths myself, but I think this is the best workaround till fix.

**Comment by sean-mcmanus** [comment]
The fix is available with 1.14.0 (pre-release): https://github.com/microsoft/vscode-cpptools/releases/tag/v1.14.0
